core/templates/services/math-interactions.service.ts

```diff
@@ -183,6 +183,100 @@ export class MathInteractionsService {
+    let bracketBalance: number = 0;
+    let shouldModifyTerm: boolean = false;
+    let modifyTerm = function(termString: string): string {
+      // If the shouldModifyTerm flag is set to true, we add the '-' sign,
+      // raise the term to a power of -1. This ensures that when the final
+      // list is joined by the '+'/'*' sign, it matches with the original
+      // expression. For eg.
+      // '3/10' would be split as [3, 10^(-1)] and
+      // '3-10' would be split as [3, -(10)].
+      if (splitByAddition) {
+        return '-(' + termString + ')';
+      } else {
+        return '(' + termString + ')^(-1)';
+      }
+    };
+    const delimiters: Array<string> = splitByAddition ? ['+', '-'] : ['*', '/'];
+
+    expressionString = expressionString.split(' ').join('');
+
+    // Temporarily replacing all unary negation signs with '~' so as to avoid
+    // splitting terms by them. We only need to split terms by binary
+    // subtraction signs and not unary negation signs. A '-' sign is considered
+    // to be binary subtraction iff it is preceded by an alphanumeric or a
+    // closing bracket, otherwise it is considered as unary negation operation.
+    // NOTE: The replace function is called twice to deal with cases where there
+    // might be overlapping matches.
+    // For eg. 4----5 would be converted to 4-~-~5 after the first call. So we
+    // need a second call to convert it to the desired result, which is 4-~~~5.
+    expressionString = expressionString.replace(/([^a-zA-Z0-9\)])-/g, '$1~');
+    expressionString = expressionString.replace(/([^a-zA-Z0-9\)])-/g, '$1~');
+
+    for (let i = 0; i < expressionString.length; i++) {
+      let currentVal = expressionString[i];
+      if (currentVal === '(' || currentVal === ')') {
+        bracketBalance += (currentVal === '(') ? 1 : -1;
+      }
+
+      // Split term only if we are not inside a set of parens and the current
+      // value is a delimiter.
+      if (bracketBalance === 0 && delimiters.indexOf(currentVal) !== -1) {
+        if (currentTerm.length !== 0) {
+          if (shouldModifyTerm) {
+            currentTerm = modifyTerm(currentTerm);
+            shouldModifyTerm = false;
+          }
+          listOfTerms.push(currentTerm);
+          currentTerm = '';
+        }
+        if (currentVal === delimiters[1]) {
+          shouldModifyTerm = true;
+        }
+      } else {
+        currentTerm += currentVal;
+      }
+    }
+    if (shouldModifyTerm) {
+      currentTerm = modifyTerm(currentTerm);
+      shouldModifyTerm = false;
+    }
+    listOfTerms.push(currentTerm);
+
+    // Reverting the temporary '~' replace in the final list of terms.
+    for (let i = 0; i < listOfTerms.length; i++) {
+      listOfTerms[i] = listOfTerms[i].replace(/~/g, '-');
+    }
+    return listOfTerms;
+  }
+
+  termsMatch(term1: string, term2: string): boolean {
+    // We split both terms by multiplication and division and try to match terms
```

Comment:
Why not addition/subtraction also? Like, terms matching `2+3` and `3+2` should also be there, right?

Reply:
Basically, this is how the entire process pans out:
1. The given expression is split into "terms". By [definition](https://simple.wikipedia.org/wiki/Term_(mathematics)), a term is a part of the expression split by + or - signs.
2. Now we will try to match terms between the learner's answer and the creator's input. These terms may internally contain * and / signs. They may also contain + or - signs but in that case, they would have to be wrapped around parens otherwise they would've been split during step 1. To match terms, we split the term into separate parts by splitting by * and /, and then performing the part match just like the term match. The split based on * and / is done in order to be able to match `3*4` with `4*3` (considering the fact that multiplication is commutative).

For eg.  consider the following scenario:
_Creator's input_: `3 + 4*5`.
_Learner's answer_: `5*4 + 3`.
**Step 1**: We split by +/- and get a list of _terms_ `[3, 4*5]` and `[5*4, 3]`.
**Step 2**: We try to match terms between them. For `4*5` and `5*4`, they get split by * or / into `[4, 5]` and `[5, 4]`. Then these lists of _parts_ gets matched just like the list of _terms_ (trying to match each term from one list with another. Note that here, they would just be checked for mathematical equivalence). And then obviously the 3s would be matched with each other directly.

Hope that makes things clearer. Please let me know if there are any more ambiguities.